proposals/XXXX-spirv-resource-representation.md

```diff
@@ -0,0 +1,239 @@
+<!-- {% raw %} -->
+
+# HLSL resources in SPIR-V
+
+*   Proposal: [NNNN](NNNN-spirv-resource-representation.md)
+*   Author(s): [Steven Perron](https://github.com/s-perron)
+*   Status: **Design In Progress**
+
+*During the review process, add the following fields as needed:*
+
+*   PRs: [#114273](https://github.com/llvm/llvm-project/pull/114273),
+    [#111052](https://github.com/llvm/llvm-project/pull/111052),
+    [#111564](https://github.com/llvm/llvm-project/pull/111564),
+    [#115178](https://github.com/llvm/llvm-project/pull/115178)
+
+## Introduction
+
+There is a need to represent the HLSL resources in llvm-ir in a way that the
+SPIR-V backend is able to create the correct code. We have already done some
+implementation work for `Buffer` and `RWBuffer`. This was done as a
+proof-of-concept, and now we needed to determine how the other resource types
+will be represented.
+
+## Motivation
+
+The HLSL resources are fundamental to HLSL, and they are required in a Vulkan
+implementation.
+
+## Proposed solution
+
+We want to match the general solution proposed in
+[0006-resource-representations.md](0006-resource-representations.md). The
+`@llvm.spv.handle.fromBinding` intrinsic will be used to get a handle to the
+resource. It will return a target type to represent the handle. Then other
+intrinsics will be used to access the resource using the handle. Previous
+proposals left open what the target types should be for SPIR-V.
+
+The general pattern for the solution will be that `@llvm.spv.handle.fromBinding`
+will return a SPIR-V pointer to a type `T`. The type for `T` will be detailed
+when discussing each HLSL resource type. The SPIR-V backend will create a global
+variable with type `T` if `range_size` is 1, and `T[range_size]` otherwise.
+
+The reason we want `@llvm.spv.handle.fromBinding` to return a pointer is to make
+it easier to satisfy the SPIR-V requirements in the
+[Universal Validation Rules](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#_universal_validation_rules).
+Specifically,
```

Comment:
I want to redo this. Types for the image types will not be pointer. It fits better with the idea that the function returns a handle. We will still use a pointer to RWStructuredBuffer and the like. In SPIR-V there is no handle for these resources. In this sense, the pointer becomes the handle.

The issue with adding the load of the handle in the same basic block has been handled by expanding postponing the expansion of `handle.fromBinding` until we see the use of the handle.